flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/assigner/EventTimeAlignedAssigner.java

```diff
@@ -0,0 +1,301 @@
+    checkAndNotifyListener();
+    return result;
+  }
+
+  private boolean hasNext() {
+    return getNextInternal().status().equals(GetSplitResult.Status.AVAILABLE);
+  }
+
+  private void checkAndNotifyListener() {
+    if (hasNext()) {
+      log.info("Looks like the future can be completed with an assignment; completing the future");
+      notifyListener();
+    }
+  }
+
+  private void notifyListener() {
+    // Simply complete the future and return;
+    futureNotifier.notifyComplete();
+  }
+
+  @Override
+  public int pendingSplitCount() {
+    return state.getUnassignedSplits().size();
+  }
+
+  @Override
+  public void onWatermarkChange(Long watermark) {
+    notifyListener();
+  }
+
+  // checks if the assigner state has reached the terminal condition
+  private boolean isTerminal() {
+    // check if there are no more splits to be added by the system
+    // check if all the splits have been completed
+    return (noMoreSplits && state.getUnassignedSplits().isEmpty());
+  }
+
+  public static class Options implements Serializable {
+
+    private final long thresholdInMs;
+
+    public Options(Duration maxMisalignmentThreshold) {
+      this.thresholdInMs = maxMisalignmentThreshold.toMillis();
+    }
+
+    public long getThresholdInMs() {
+      return thresholdInMs;
+    }
+  }
+
+  /**
+   * Keeps track of unassigned splits ordered by certain comparator.
+   */
+  @NotThreadSafe
+  static
+  class State {
+    private static final Logger log = LoggerFactory.getLogger(State.class);
+
+    private final WatermarkTracker tracker;
+
+    private final TreeSet<IcebergSourceSplit> unassignedSplits;
+    private final TreeSet<IcebergSourceSplit> assignedSplits;
+    private final TimestampAssigner<IcebergSourceSplit> timestampAssigner;
+
+    public State(
+        WatermarkTracker tracker,
+        TimestampAssigner<IcebergSourceSplit> timestampAssigner) {
+      this.tracker = tracker;
+      this.timestampAssigner = timestampAssigner;
+      Comparator<IcebergSourceSplit> comparator =
+          new AscendingTimestampSplitComparator(timestampAssigner);
+      this.unassignedSplits = new TreeSet<>(comparator);
+      this.assignedSplits = new TreeSet<>(comparator);
+    }
+
+    public Collection<IcebergSourceSplit> getUnassignedSplits() {
+      return Collections.unmodifiableCollection(unassignedSplits);
+    }
+
+    void onDiscoveredSplits(Collection<IcebergSourceSplit> splits) {
+      unassignedSplits.addAll(splits);
+      updateWatermark();
+    }
+
+    public void onSplitAssigned(IcebergSourceSplit split) {
+      if (!unassignedSplits.contains(split)) {
+        log.error("split={} not found in unassigned splits", split);
+        throw new IllegalArgumentException("split not found in unassigned splits");
+      }
+      unassignedSplits.remove(split);
+      assignedSplits.add(split);
+      updateWatermark();
+    }
+
+    public void onUnassignedSplits(Collection<IcebergSourceSplit> splits) {
```

Comment:
Are we sure that the unassigned splits are not causing issues with other TaskManagers watermarks?

Can the situation below happen, or I miss something?
1. We have Splits: S1, S2, S3, S4 - in order
2. TM1 (TaskManager1) got a S1 assigned
3. TM2 got S2 assigned
4. TM1 fails - the S1 is unassigned
5. TM2 finishes S2 and gets S1 - in this case S1 is out of order and might cause issues 

We might be good if the S1 would get assigned to the restarted TM1', but if it is assigned to TM2 then it is out of order.

Reply:
Ohh.. I think I understand it. That is why we customize the watermarks. We do not advance the watermarks until a split is not finished.

That would mean that `updateWatermark` here is not necessary, as we should not receive splits  here which are not already in the `assignedSplits` set. Maybe we should check that, otherwise if we decrease a watermark, then we will have an issue.